lstchain/reco/reconstructor.py

```diff
@@ -0,0 +1,944 @@
+        calibrated response of the pixels.
+
+    """
+
+    def __init__(self, *args, n_peaks=100, **kwargs):
+        """
+            Initialise the data and parameters used for the fitting, including
+            method specific objects.
+
+            Parameters:
+            -----------
+            *args, **kwargs: Argument for the DL0Fitter initialisation
+            n_peak: int
+                Maximum upper bound of the sum over possible detected
+                photo-electron value in the likelihood computation.
+
+        """
+        super().__init__(*args, **kwargs)
+        self._initialize_pdf(n_peaks=n_peaks)
+
+    def clean_data(self):
+        """
+            Method used to select pixels and time samples used in the
+            fitting procedure. The spatial selection takes pixels in an
+            ellipsis obtained from the seed Hillas parameters extraction.
+            An additional factor sigma_space is added to the length and width
+            to extend the selected region. The temporal selection takes a time
+            window centered on the seed time of center of mass and of duration
+            equal to the time of propagation of the signal along the length
+            of the ellipsis times a factor sigma_time. An additional fixed
+            duration is also added before and after this time window through
+            the time_before_shower and time_after_shower arguments.
+
+        """
+        x_cm = self.start_parameters['x_cm']
+        y_cm = self.start_parameters['y_cm']
+        length = self.start_parameters['length']
+        width = self.start_parameters['wl'] * length
+        psi = self.start_parameters['psi']
+
+        dx = self.pix_x - x_cm
+        dy = self.pix_y - y_cm
+
+        lon = dx * np.cos(psi) + dy * np.sin(psi)
+        lat = dx * np.sin(psi) - dy * np.cos(psi)
+
+        mask_pixel = ((lon / (length+0.0228)) ** 2 + (
+                    lat / (width+0.0228)) ** 2) < self.sigma_space ** 2
+
+        v = self.start_parameters['v']
+        t_start = (self.start_parameters['t_cm']
+                   - (np.abs(v) * length / 2 * self.sigma_time)
+                   - self.time_before_shower)
+        t_end = (self.start_parameters['t_cm']
+                 + (np.abs(v) * length / 2 * self.sigma_time)
+                 + self.time_after_shower)
+
+        mask_time = (self.times < t_end) * (self.times > t_start)
+
+        return mask_pixel, mask_time
+
+    def _initialize_pdf(self, n_peaks):
+        """
+            Compute quantities used at each iteration of the fitting procedure.
+        """
+        self.n_peaks = n_peaks
+        photoelectron_peak = np.arange(n_peaks, dtype=int)
+        photoelectron_peak[0] = 1
+        log_factorial = np.log(photoelectron_peak)
+        log_factorial = np.cumsum(log_factorial)
+        self.log_factorial = log_factorial
+
+    def log_pdf(self, charge, t_cm, x_cm, y_cm,
```

Comment:
Looking at this, I am not surprised this is very slow. This function needs to be compiled. Using numba would probably be easiest.